crates/bevy_ecs/src/entity/clone_entities.rs

```diff
@@ -1,20 +1,225 @@
     entity::Entity,
     world::World,
 };
 
-/// A helper struct to clone an entity. Used internally by [`EntityCloneBuilder::clone_entity`] and custom clone handlers.
+/// Context for component clone handlers.
+///
+/// Provides fast access to useful resources like [`AppTypeRegistry`](crate::reflect::AppTypeRegistry)
+/// and allows component clone handler to get information about component being cloned.
+pub struct ComponentCloneCtx<'a, 'b> {
+    component_id: ComponentId,
+    source_component_ptr: Ptr<'a>,
+    target_component_written: bool,
+    target_components_ptrs: &'a mut Vec<PtrMut<'b>>,
+    target_components_buffer: &'b Bump,
+    components: &'a Components,
+    entity_cloner: &'a EntityCloner,
+    #[cfg(feature = "bevy_reflect")]
+    type_registry: Option<&'a crate::reflect::AppTypeRegistry>,
+}
+
+impl<'a, 'b> ComponentCloneCtx<'a, 'b> {
+    /// Create a new instance of `ComponentCloneCtx` that can be passed to component clone handlers.
+    ///
+    /// # Safety
+    /// Caller must ensure that:
+    /// - `components` and `component_id` are from the same world.
+    /// - `source_component_ptr` points to a valid component of type represented by `component_id`.
+    unsafe fn new(
+        component_id: ComponentId,
+        source_component_ptr: Ptr<'a>,
+        target_components_ptrs: &'a mut Vec<PtrMut<'b>>,
+        target_components_buffer: &'b Bump,
+        components: &'a Components,
+        entity_cloner: &'a EntityCloner,
+        #[cfg(feature = "bevy_reflect")] type_registry: Option<&'a crate::reflect::AppTypeRegistry>,
+    ) -> Self {
+        Self {
+            component_id,
+            source_component_ptr,
+            target_components_ptrs,
+            target_component_written: false,
+            target_components_buffer,
+            components,
+            entity_cloner,
+            #[cfg(feature = "bevy_reflect")]
+            type_registry,
+        }
+    }
+
+    /// Returns true if [`write_target_component`](`Self::write_target_component`) was called before.
+    pub fn target_component_written(&self) -> bool {
+        self.target_component_written
+    }
+
+    /// Returns the current source entity.
+    pub fn source(&self) -> Entity {
+        self.entity_cloner.source
+    }
+
+    /// Returns the current target entity.
+    pub fn target(&self) -> Entity {
+        self.entity_cloner.target
+    }
+
+    /// Returns the [`ComponentId`] of currently cloned component.
+    pub fn component_id(&self) -> ComponentId {
+        self.component_id
+    }
+
+    /// Returns a reference to the component on the source entity.
+    ///
+    /// Will return `None` if `ComponentId` of requested component does not match `ComponentId` of source component
+    pub fn read_source_component<T: Component>(&self) -> Option<&T> {
+        if self
+            .components
+            .component_id::<T>()
+            .is_some_and(|id| id == self.component_id)
```

Comment:
> This also means it can't work with dynamic components.

I think that we can probably allow to clone dynamic components by exposing some unsafe methods on `ComponentCloneCtx` to get source component `Ptr` and to copy from provided `Ptr` to target. This would let someone who created dynamic component to register their own component clone handler.